Job Post Analysis and Report

In [1]:
# Import General Libraries
import pandas as pd 
import numpy as np 

# Import Visualization Libraries
import matplotlib.pyplot as plt 
import seaborn as sns 

## Loading Data

In [2]:
df = pd.read_csv(r'transformed_jobs_data.csv')
df.head(3)

job_id job_employment_type  \
0  XX6946dvNO3187IkAAAAAA==            FULLTIME   
1  6Wu7QcWfrhOR-THoAAAAAA==            FULLTIME   
2  Oiu186OT5E6bkZReAAAAAA==            FULLTIME   

                                           job_title  \
0                                      Data Engineer   
1                                      Data Engineer   
2  2024 RBCIS, Summer Co-op Data Engineer / Full ...   

                                      job_apply_link  \
0  https://ca.linkedin.com/jobs/view/data-enginee...   
1  https://ca.linkedin.com/jobs/view/data-enginee...   
2  https://ca.linkedin.com/jobs/view/2024-rbcis-s...   

                                     job_description  job_city job_country  \
0  • *Please note that our offices will be closed...  Montréal          CA   
1  We are seeking an experienced and passionate D...       NaN          CA   
2  Job Summary\n\nJob Description\n\nWhat is the ...   Toronto          CA   

  job_posted_at_timestamp    employer_website employer_company_type  
0     2023-12-19 20:22:04                 NaN                   NaN  
1     2023-12-19 15:55:49                 NaN                   NaN  
2     2023-12-19 22:30:06  http://www.rbc.com               Finance

In [3]:
# df['job_city'].unique()

In [4]:
df.tail()

job_id job_employment_type                   job_title  \
161  0JxLam0cJtXsmjQZAAAAAA==            FULLTIME  Senior Cloud Data Engineer   
162  EGvoxwF7K6W0LOMtAAAAAA==            FULLTIME  Senior Cloud Data Engineer   
163  TvhAbNPDfD24fyHVAAAAAA==            FULLTIME  Senior Cloud Data Engineer   
164  6N_3eAHv_8HXYM0cAAAAAA==            FULLTIME  Senior Cloud Data Engineer   
165  iQu2KNPFHbJstSZuAAAAAA==            FULLTIME  Senior Cloud Data Engineer   

                                        job_apply_link  \
161  https://www.linkedin.com/jobs/view/senior-clou...   
162  https://www.linkedin.com/jobs/view/senior-clou...   
163  https://www.linkedin.com/jobs/view/senior-clou...   
164  https://www.linkedin.com/jobs/view/senior-clou...   
165  https://www.linkedin.com/jobs/view/senior-clou...   

                                       job_description         job_city  \
161  Job Description\n\nJob Summary:\n\nThis positi...        Las Vegas   
162  Job Description\n\nJob Summary:\n\nThis positi...         Richmond   
163  Job Description\n\nJob Summary:\n\nThis positi...       Fort Worth   
164  Job Description\n\nJob Summary:\n\nThis positi...         Columbus   
165  Job Description\n\nJob Summary:\n\nThis positi...  Fort Lauderdale   

    job_country job_posted_at_timestamp    employer_website  \
161          US     2023-12-19 12:11:14  http://www.bdo.com   
162          US     2023-12-19 10:42:36  http://www.bdo.com   
163          US     2023-12-19 11:52:55  http://www.bdo.com   
164          US     2023-12-19 12:15:04  http://www.bdo.com   
165          US     2023-12-19 12:11:14  http://www.bdo.com   

    employer_company_type  
161            Consulting  
162            Consulting  
163            Consulting  
164            Consulting  
165            Consulting

In [5]:
df.shape

(166, 10)

In [6]:
df.isnull().sum()

job_id                      0
job_employment_type         0
job_title                   0
job_apply_link              0
job_description             0
job_city                   38
job_country                 0
job_posted_at_timestamp     0
employer_website           53
employer_company_type      90
dtype: int64

In [7]:
# Function to format null values to 2dp
def formatt(x):
    return '{:.2f}%'.format(x)

((df.isnull().sum() / len(df)) * 100).apply(formatt)

job_id                      0.00%
job_employment_type         0.00%
job_title                   0.00%
job_apply_link              0.00%
job_description             0.00%
job_city                   22.89%
job_country                 0.00%
job_posted_at_timestamp     0.00%
employer_website           31.93%
employer_company_type      54.22%
dtype: object

- Presence of missing data in **THREE** columns namely:
    - job_city [22.89%]
    - employer_website [31.93%]
    - employer_company_type [54.22%]

In [8]:
df.nunique()

job_id                     166
job_employment_type          4
job_title                   99
job_apply_link             166
job_description            153
job_city                    70
job_country                  3
job_posted_at_timestamp    154
employer_website            88
employer_company_type       13
dtype: int64

In [9]:
df.duplicated().value_counts()

False    166
Name: count, dtype: int64

- No duplicates.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166 entries, 0 to 165
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   job_id                   166 non-null    object
 1   job_employment_type      166 non-null    object
 2   job_title                166 non-null    object
 3   job_apply_link           166 non-null    object
 4   job_description          166 non-null    object
 5   job_city                 128 non-null    object
 6   job_country              166 non-null    object
 7   job_posted_at_timestamp  166 non-null    object
 8   employer_website         113 non-null    object
 9   employer_company_type    76 non-null     object
dtypes: object(10)
memory usage: 13.1+ KB


In [11]:
# Impute missing job_city 
modeCity = df.groupby('job_country')['job_city'].count().reset_index()
modeCity.columns = ['job_country', 'job_city']
modeCity

    

job_country  job_city
0          CA        26
1          GB        48
2          US        54

In [30]:
df.dtypes

job_id                     object
job_employment_type        object
job_title                  object
job_apply_link             object
job_description            object
job_city                   object
job_country                object
job_posted_at_timestamp    object
employer_website           object
employer_company_type      object
dtype: object

In [31]:
# Convert job_city to a categorical data
df['job_city'] = df['job_city'].astype('category')

In [35]:
# Label Encoding in Pandas
df['Encoding_with_pandas'] = df['job_city'].cat.codes
df.head(3)

job_id job_employment_type  \
0  XX6946dvNO3187IkAAAAAA==            FULLTIME   
1  6Wu7QcWfrhOR-THoAAAAAA==            FULLTIME   
2  Oiu186OT5E6bkZReAAAAAA==            FULLTIME   

                                           job_title  \
0                                      Data Engineer   
1                                      Data Engineer   
2  2024 RBCIS, Summer Co-op Data Engineer / Full ...   

                                      job_apply_link  \
0  https://ca.linkedin.com/jobs/view/data-enginee...   
1  https://ca.linkedin.com/jobs/view/data-enginee...   
2  https://ca.linkedin.com/jobs/view/2024-rbcis-s...   

                                     job_description  job_city job_country  \
0  • *Please note that our offices will be closed...  Montréal          CA   
1  We are seeking an experienced and passionate D...       NaN          CA   
2  Job Summary\n\nJob Description\n\nWhat is the ...   Toronto          CA   

  job_posted_at_timestamp    employer_website employer_company_type  \
0     2023-12-19 20:22:04                 NaN                   NaN   
1     2023-12-19 15:55:49                 NaN                   NaN   
2     2023-12-19 22:30:06  http://www.rbc.com               Finance   

   Encoding_with_pandas  
0                    44  
1                    -1  
2                    64

In [37]:
# Label encoding using Scikit-Learn
from sklearn.preprocessing import LabelEncoder

# Create an instance
label_encoder = LabelEncoder()

# Fit LabelEncoder
label_encoder.fit(df['job_city'])

df['encoded_with_sklearn'] = label_encoder.transform(df['job_city'])

In [38]:
df.head()

job_id job_employment_type  \
0  XX6946dvNO3187IkAAAAAA==            FULLTIME   
1  6Wu7QcWfrhOR-THoAAAAAA==            FULLTIME   
2  Oiu186OT5E6bkZReAAAAAA==            FULLTIME   
3  5y7azm6gLeu1ab_9AAAAAA==            FULLTIME   
4  CJUJsgmqVdDmJmHgAAAAAA==            FULLTIME   

                                           job_title  \
0                                      Data Engineer   
1                                      Data Engineer   
2  2024 RBCIS, Summer Co-op Data Engineer / Full ...   
3                                 Lead Data Engineer   
4  2024 Investor Services - Business Data Analyst...   

                                      job_apply_link  \
0  https://ca.linkedin.com/jobs/view/data-enginee...   
1  https://ca.linkedin.com/jobs/view/data-enginee...   
2  https://ca.linkedin.com/jobs/view/2024-rbcis-s...   
3  https://ca.linkedin.com/jobs/view/lead-data-en...   
4  https://jobs.rbc.com/ca/en/job/R-0000074144/20...   

                                     job_description  job_city job_country  \
0  • *Please note that our offices will be closed...  Montréal          CA   
1  We are seeking an experienced and passionate D...       NaN          CA   
2  Job Summary\n\nJob Description\n\nWhat is the ...   Toronto          CA   
3  Overview\n\nAt Nortal, we are driven by a gran...       NaN          CA   
4  Job Summary\n\nWhat is the opportunity?\n\nAs ...   Toronto          CA   

  job_posted_at_timestamp       employer_website employer_company_type  \
0     2023-12-19 20:22:04                    NaN                   NaN   
1     2023-12-19 15:55:49                    NaN                   NaN   
2     2023-12-19 22:30:06     http://www.rbc.com               Finance   
3     2023-12-19 17:06:15  http://www.nortal.com     Computer Services   
4     2023-12-20 00:00:00     http://www.rbc.com               Finance   

   Encoding_with_pandas  encoded_with_sklearn  
0                    44                    44  
1                    -1                    70  
2                    64                    64  
3                    -1                    70  
4                    64                    64

In [44]:
# One Hot Encoding
from sklearn.preprocessing import OneHotEncoder

one_hot_encoder = OneHotEncoder()

transformed_one_hot_encoding = one_hot_encoder.fit_transform(df[['job_city']])

ohe_df = pd.DataFrame(transformed_one_hot_encoding.toarray(), dtype='int')
ohe_df.head()

0   1   2   3   4   5   6   7   8   9   ...  61  62  63  64  65  66  67  \
0   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
1   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
2   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   1   0   0   0   
3   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
4   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   1   0   0   0   

   68  69  70  
0   0   0   0  
1   0   0   1  
2   0   0   0  
3   0   0   1  
4   0   0   0  

[5 rows x 71 columns]

In [45]:
ohe_final_df = df.join(ohe_df)
ohe_final_df

job_id job_employment_type  \
0    XX6946dvNO3187IkAAAAAA==            FULLTIME   
1    6Wu7QcWfrhOR-THoAAAAAA==            FULLTIME   
2    Oiu186OT5E6bkZReAAAAAA==            FULLTIME   
3    5y7azm6gLeu1ab_9AAAAAA==            FULLTIME   
4    CJUJsgmqVdDmJmHgAAAAAA==            FULLTIME   
..                        ...                 ...   
161  0JxLam0cJtXsmjQZAAAAAA==            FULLTIME   
162  EGvoxwF7K6W0LOMtAAAAAA==            FULLTIME   
163  TvhAbNPDfD24fyHVAAAAAA==            FULLTIME   
164  6N_3eAHv_8HXYM0cAAAAAA==            FULLTIME   
165  iQu2KNPFHbJstSZuAAAAAA==            FULLTIME   

                                             job_title  \
0                                        Data Engineer   
1                                        Data Engineer   
2    2024 RBCIS, Summer Co-op Data Engineer / Full ...   
3                                   Lead Data Engineer   
4    2024 Investor Services - Business Data Analyst...   
..                                                 ...   
161                         Senior Cloud Data Engineer   
162                         Senior Cloud Data Engineer   
163                         Senior Cloud Data Engineer   
164                         Senior Cloud Data Engineer   
165                         Senior Cloud Data Engineer   

                                        job_apply_link  \
0    https://ca.linkedin.com/jobs/view/data-enginee...   
1    https://ca.linkedin.com/jobs/view/data-enginee...   
2    https://ca.linkedin.com/jobs/view/2024-rbcis-s...   
3    https://ca.linkedin.com/jobs/view/lead-data-en...   
4    https://jobs.rbc.com/ca/en/job/R-0000074144/20...   
..                                                 ...   
161  https://www.linkedin.com/jobs/view/senior-clou...   
162  https://www.linkedin.com/jobs/view/senior-clou...   
163  https://www.linkedin.com/jobs/view/senior-clou...   
164  https://www.linkedin.com/jobs/view/senior-clou...   
165  https://www.linkedin.com/jobs/view/senior-clou...   

                                       job_description         job_city  \
0    • *Please note that our offices will be closed...         Montréal   
1    We are seeking an experienced and passionate D...              NaN   
2    Job Summary\n\nJob Description\n\nWhat is the ...          Toronto   
3    Overview\n\nAt Nortal, we are driven by a gran...              NaN   
4    Job Summary\n\nWhat is the opportunity?\n\nAs ...          Toronto   
..                                                 ...              ...   
161  Job Description\n\nJob Summary:\n\nThis positi...        Las Vegas   
162  Job Description\n\nJob Summary:\n\nThis positi...         Richmond   
163  Job Description\n\nJob Summary:\n\nThis positi...       Fort Worth   
164  Job Description\n\nJob Summary:\n\nThis positi...         Columbus   
165  Job Description\n\nJob Summary:\n\nThis positi...  Fort Lauderdale   

    job_country job_posted_at_timestamp       employer_website  \
0            CA     2023-12-19 20:22:04                    NaN   
1            CA     2023-12-19 15:55:49                    NaN   
2            CA     2023-12-19 22:30:06     http://www.rbc.com   
3            CA     2023-12-19 17:06:15  http://www.nortal.com   
4            CA     2023-12-20 00:00:00     http://www.rbc.com   
..          ...                     ...                    ...   
161          US     2023-12-19 12:11:14     http://www.bdo.com   
162          US     2023-12-19 10:42:36     http://www.bdo.com   
163          US     2023-12-19 11:52:55     http://www.bdo.com   
164          US     2023-12-19 12:15:04     http://www.bdo.com   
165          US     2023-12-19 12:11:14     http://www.bdo.com   

    employer_company_type  ...  61  62  63  64  65  66  67  68  69  70  
0                     NaN  ...   0   0   0   0   0   0   0   0   0   0  
1                     NaN  ...   0   0   0   0   0   0   0   0   0   1  
2                 Finance  ...   0   0   0   1   0   0   0   0   0   0  
3     

In [48]:
# Generate OHE using pandas dummy
dummy_df = pd.get_dummies(df['job_city'], prefix='job_city')
dummy_df

job_city_Ajax  job_city_Atlanta  job_city_Auburn Hills  job_city_Austin  \
0            False             False                  False            False   
1            False             False                  False            False   
2            False             False                  False            False   
3            False             False                  False            False   
4            False             False                  False            False   
..             ...               ...                    ...              ...   
161          False             False                  False            False   
162          False             False                  False            False   
163          False             False                  False            False   
164          False             False                  False            False   
165          False             False                  False            False   

     job_city_Baltimore  job_city_Belfast  job_city_Birmingham  \
0                 False             False                False   
1                 False             False                False   
2                 False             False                False   
3                 False             False                False   
4                 False             False                False   
..                  ...               ...                  ...   
161               False             False                False   
162               False             False                False   
163               False             False                False   
164               False             False                False   
165               False             False                False   

     job_city_Boston  job_city_Bristol  job_city_Cambridge  ...  \
0              False             False               False  ...   
1              False             False               False  ...   
2              False             False               False  ...   
3              False             False               False  ...   
4              False             False               False  ...   
..               ...               ...                 ...  ...   
161            False             False               False  ...   
162            False             False               False  ...   
163            False             False               False  ...   
164            False             False               False  ...   
165            False             False               False  ...   

     job_city_Rolling Meadows  job_city_San Diego  job_city_Seattle  \
0                       False               False             False   
1                       False               False             False   
2                       False               False             False   
3                       False               False             False   
4                       False               False             False   
..                        ...                 ...               ...   
161                     False               False             False   
162                     False               False             False   
163                     False               False             False   
164                     False               False             False   
165                     False               False             False   

     job_city_Springdale  job_city_Toronto  job_city_Vancouver  \
0                  False             False               False   
1                  False             False               False   
2                  False              True               False   
3                  False             False               False   
4                  False              True               False   
..                   ...               ...                 ...   
161                False             False               False   
162                False             False               False   

In [49]:
dummy_df.style.applymap(lambda x: 'background-color: green' if x > 0 else 'background-color: yellow')

In [12]:
# Function to fill missing volumes with the mode value of each city
# def fill_missing_city(df):
#     if np.isnan(df['job_city']):
#         return modeCity[modeCity['job_country'] == df['job_country']]['mode']
    
#     return df['job_city']

# df['job_city'] = df.apply(fill_missing_city, axis=1)

In [13]:
df.columns

Index(['job_id', 'job_employment_type', 'job_title', 'job_apply_link',
       'job_description', 'job_city', 'job_country', 'job_posted_at_timestamp',
       'employer_website', 'employer_company_type'],
      dtype='object')

In [16]:
len(df['job_city'].unique())

71

In [ ]:
# imputer = SimpleImputer(strategy='most_frequent')
# imputer.fit_transform(df['job_city'])

In [ ]:
# Impute job_city using the most frequent value
# from sklearn.impute import SimpleImputer
# impute_mode = SimpleImputer(strategy='most_frequent')
# impute_mode.fit(df[['job_city']])

# df[['job_city']] = impute_mode.fit_transform(df[['job_city']].values.reshape(-1,1))
# df['job_city'].isna().sum()


0

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166 entries, 0 to 165
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   job_id                   166 non-null    object
 1   job_employment_type      166 non-null    object
 2   job_title                166 non-null    object
 3   job_apply_link           166 non-null    object
 4   job_description          166 non-null    object
 5   job_city                 128 non-null    object
 6   job_country              166 non-null    object
 7   job_posted_at_timestamp  166 non-null    object
 8   employer_website         113 non-null    object
 9   employer_company_type    76 non-null     object
dtypes: object(10)
memory usage: 13.1+ KB


In [ ]:
# Drop missing rows
# df_copy = df.copy()

# df_copy.dropna(inplace=True)

# df_copy.isna().sum()

### Exploratory Data Analysis